modules/frontend/frontend.go

```diff
@@ -0,0 +1,227 @@
+		Namespace: "tempo",
+		Name:      "query_frontend_queries_total",
+		Help:      "Total queries sent per tenant.",
+	}, []string{"user"})
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		// get the http request, add some custom parameters to it, split it, and call downstream roundtripper
+		rt := NewRoundTripper(next, ShardingWare(cfg.ShardNum, logger, registerer))
+		return frontend.RoundTripFunc(func(r *http.Request) (*http.Response, error) {
+			level.Info(util.Logger).Log("msg", "request received by custom tripperware")
+			orgID := r.Header.Get(user.OrgIDHeaderName)
+			queriesPerTenant.WithLabelValues(orgID).Inc()
+
+			r = r.WithContext(user.InjectOrgID(r.Context(), orgID))
+			return rt.RoundTrip(r)
+		})
+	}, nil
+}
+
+type Handler interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+type Middleware interface {
+	Wrap(Handler) Handler
+}
+
+// MiddlewareFunc is like http.HandlerFunc, but for Middleware.
+type MiddlewareFunc func(Handler) Handler
+
+// Wrap implements Middleware.
+func (q MiddlewareFunc) Wrap(h Handler) Handler {
+	return q(h)
+}
+
+func MergeMiddlewares(middleware ...Middleware) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		for i := len(middleware) - 1; i >= 0; i-- {
+			next = middleware[i].Wrap(next)
+		}
+		return next
+	})
+}
+
+type roundTripper struct {
+	next    http.RoundTripper
+	handler Handler
+}
+
+// NewRoundTripper merges a set of middlewares into an handler, then inject it into the `next` roundtripper
+func NewRoundTripper(next http.RoundTripper, middlewares ...Middleware) http.RoundTripper {
+	transport := roundTripper{
+		next:  next,
+	}
+	transport.handler = MergeMiddlewares(middlewares...).Wrap(&transport)
+	return transport
+}
+
+func (q roundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	level.Info(util.Logger).Log("blockStart", r.URL.Query().Get("blockStart"), "blockEnd", r.URL.Query().Get("blockEnd"))
+	return q.handler.Do(r)
+}
+
+// Do implements Handler.
+func (q roundTripper) Do(r *http.Request) (*http.Response, error) {
+	level.Info(util.Logger).Log("msg", "roundTripper.Do called")
+	return q.next.RoundTrip(r)
+}
+
+func ShardingWare(shardNum int, logger log.Logger, registerer prometheus.Registerer) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return shardQuery{
+			next:     next,
+			shardNum: shardNum,
+			logger:   logger,
+			splitByCounter: promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+				Namespace: "tempo",
+				Name:      "query_frontend_split_queries_total",
+				Help:      "Total number of underlying query requests after sharding",
+			}, []string{"user"}),
+		}
+	})
+}
+
+type shardQuery struct {
+	next     Handler
+	shardNum int
+	logger   log.Logger
+	// Metrics.
+	splitByCounter *prometheus.CounterVec
+}
+
+// Do implements Handler
+func (s shardQuery) Do(r *http.Request) (*http.Response, error) {
```

Comment:
this method needs to be tested.  if easier segregating and just testing the code that shards the block space is fine.